src/eos-update-server.c

```diff
@@ -0,0 +1,863 @@
+{
+  gconstpointer raw;
+  gsize len;
+
+  raw = g_bytes_get_data (bytes, &len);
+  return soup_buffer_new_with_owner (raw,
+                                     len,
+                                     g_bytes_ref (bytes),
+                                     (GDestroyNotify)g_bytes_unref);
+}
+
+static gboolean
+send_bytes (SoupMessage *msg,
+            GBytes *bytes)
+{
+  g_autoptr(SoupBuffer) buffer = buffer_from_bytes (bytes);
+
+  soup_message_body_append_buffer (msg->response_body, buffer);
+  soup_message_set_status (msg, SOUP_STATUS_OK);
+
+  return TRUE;
+}
+
+static gboolean
+handle_config (SoupMessage *msg,
+               Data *data)
+{
+  return send_bytes (msg, data->cached_config);
+}
+
+static gboolean
+serve_file_if_exists (SoupMessage *msg,
+                      const gchar *raw_path,
+                      gboolean *served)
+{
+  g_autoptr(GFile) path = g_file_new_for_path (raw_path);
+  g_autoptr(GMappedFile) mapping = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(SoupBuffer) buffer = NULL;
+
+  if (!g_file_query_exists (path, NULL))
+    {
+      *served = FALSE;
+      return TRUE;
+    }
+
+  mapping = g_mapped_file_new (raw_path, FALSE, &error);
+  if (mapping == NULL)
+    {
+      message ("Failed to map %s: %s", raw_path, error->message);
+      soup_message_set_status (msg, SOUP_STATUS_INTERNAL_SERVER_ERROR);
+      return FALSE;
+    }
+
+  message ("Serving %s", raw_path);
+  buffer = soup_buffer_new_with_owner (g_mapped_file_get_contents (mapping),
+                                       g_mapped_file_get_length (mapping),
+                                       g_mapped_file_ref (mapping),
+                                       (GDestroyNotify)g_mapped_file_unref);
+  soup_message_body_append_buffer (msg->response_body, buffer);
+  soup_message_set_status (msg, SOUP_STATUS_OK);
+  *served = TRUE;
+
+  return TRUE;
+}
+
+static gboolean
+serve_file (SoupMessage *msg,
+            const gchar *raw_path)
+{
+  gboolean served = FALSE;
+
+  if (!serve_file_if_exists (msg, raw_path, &served))
+    return FALSE;
+
+  if (!served)
+    {
+      message ("File %s not found", raw_path);
+      soup_message_set_status (msg, SOUP_STATUS_NOT_FOUND);
+    }
+
+  return served;
+}
+
+static gboolean
+handle_refs_heads (SoupMessage *msg,
+                   Data *data,
+                   const gchar *requested_path)
+{
+  g_autofree gchar *raw_path = g_build_filename (data->cached_repo_root, requested_path, NULL);
+  gboolean served = FALSE;
+  gsize prefix_len;
+  size_t len;
+  const gchar *head;
+
+  if (!serve_file_if_exists (msg, raw_path, &served))
+    return FALSE;
+
+  if (served)
+    return TRUE;
+
+  g_clear_pointer (&raw_path, g_free);
+  prefix_len = 1 + 4 + 1 + 5 + 1; /* /refs/heads/ */
+  len = strlen (requested_path);
+  head = requested_path + prefix_len; /* e.g eos2/i386 */
+
+  if (len <= prefix_len)
+    {
+      message ("Invalid request for /refs/heads/");
```

Comment:
Is the intention of this to check whether `/refs/heads/` is a prefix of `requested_path`? Maybe it would be better to use `g_str_has_prefix`?